File: frontend/web-console/services/documents.ts
/**
 * Document ingestion service
 * 
 * Connects to ingestion-service API for file upload and management.
 */

import axios from 'axios';
import { tokenManager } from '@/core/api-client';

const INGESTION_API = process.env.NEXT_PUBLIC_INGESTION_API || 'http://localhost:8002';

// Types matching ingestion-service schemas
export interface DocumentUploadResponse {
    document_id: string;
    status: string;
    filename: string;
    content_type: string;
    size_bytes: number;
    file_category: string;
    uploaded_at: string;
}

export interface Document {
    document_id: string;
    filename: string;
    content_type: string;
    size_bytes: number;
    file_category: string;
    status: string;
    language?: string;
    retention_policy: string;
    legal_hold: boolean;
    uploaded_by: string;
    department_id?: string;
    collection_id?: string;
    uploaded_at: string;
    validated_at?: string;
    processed_at?: string;
}

export interface DocumentListItem {
    document_id: string;
    filename: string;
    content_type: string;
    size_bytes: number;
    file_category: string;
    status: string;
    uploaded_at: string;
}

export interface DocumentListResponse {
    items: DocumentListItem[];
    next_cursor?: string;
    total_count?: number;
}

export interface DeleteResponse {
    document_id: string;
    status: string;
    message: string;
}

// Create axios client for ingestion service
const ingestionClient = axios.create({
    baseURL: INGESTION_API,
    timeout: 300000, // 5 min for large uploads
    withCredentials: true,
});

// Add auth headers
ingestionClient.interceptors.request.use((config) => {
    const token = tokenManager.getAccessToken();
    const tenantId = tokenManager.getTenantId();

    if (token) {
        config.headers.Authorization = `Bearer ${token}`;
    }
    if (tenantId) {
        config.headers['X-Tenant-ID'] = tenantId;
    }

    return config;
});

/**
 * Upload a document to the ingestion service
 */
export async function uploadDocument(
    file: File,
    collectionId?: string,
    onProgress?: (percent: number) => void
): Promise<DocumentUploadResponse> {
    const formData = new FormData();
    formData.append('file', file);

    if (collectionId) {
        formData.append('collection_id', collectionId);
    }

    const response = await ingestionClient.post('/api/v1/documents', formData, {
        headers: {
            'Content-Type': 'multipart/form-data',
        },
        onUploadProgress: (progressEvent) => {
            if (onProgress && progressEvent.total) {
                const percent = Math.round((progressEvent.loaded * 100) / progressEvent.total);
                onProgress(percent);
            }
        },
    });

    return response.data;
}

/**
 * List documents with pagination and filters
 */
export async function listDocuments(params?: {
    limit?: number;
    cursor?: string;
    status?: string;
    file_category?: string;
    collection_id?: string;
}): Promise<DocumentListResponse> {
    const response = await ingestionClient.get('/api/v1/documents', {
        params: {
            limit: params?.limit ?? 20,
            cursor: params?.cursor,
            status: params?.status,
            file_category: params?.file_category,
            collection_id: params?.collection_id,
        },
    });

    return response.data;
}

/**
 * Get a single document by ID
 */
export async function getDocument(documentId: string): Promise<Document> {
    const response = await ingestionClient.get(`/api/v1/documents/${documentId}`);
    return response.data;
}

/**
 * Delete a document (soft delete)
 */
export async function deleteDocument(documentId: string): Promise<DeleteResponse> {
    const response = await ingestionClient.delete(`/api/v1/documents/${documentId}`);
    return response.data;
}

/**
 * Get file category icon
 */
export function getFileCategoryIcon(category: string): string {
    switch (category.toUpperCase()) {
        case 'DOCUMENT':
            return '📄';
        case 'IMAGE':
            return '🖼️';
        case 'AUDIO':
            return '🎵';
        case 'VIDEO':
            return '🎬';
        default:
            return '📁';
    }
}

/**
 * Format file size for display
 */
export function formatFileSize(bytes: number): string {
    if (bytes === 0) return '0 B';
    const k = 1024;
    const sizes = ['B', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return `${parseFloat((bytes / Math.pow(k, i)).toFixed(1))} ${sizes[i]}`;
}

/**
 * Allowed file types for upload
 */
export const ALLOWED_FILE_TYPES = {
    'application/pdf': { label: 'PDF', maxSize: 100 * 1024 * 1024 },
    'image/png': { label: 'PNG', maxSize: 50 * 1024 * 1024 },
    'image/jpeg': { label: 'JPEG', maxSize: 50 * 1024 * 1024 },
    'audio/mpeg': { label: 'MP3', maxSize: 500 * 1024 * 1024 },
    'audio/wav': { label: 'WAV', maxSize: 500 * 1024 * 1024 },
    'video/mp4': { label: 'MP4', maxSize: 2048 * 1024 * 1024 },
};

export function isAllowedFileType(file: File): boolean {
    return file.type in ALLOWED_FILE_TYPES;
}

export function getMaxFileSize(file: File): number {
    const config = ALLOWED_FILE_TYPES[file.type as keyof typeof ALLOWED_FILE_TYPES];
    return config?.maxSize ?? 100 * 1024 * 1024;
}

File: services/ingestion-service/src/ingestion_service/api/routes/documents.py
"""
Document API routes.

Handles upload, listing, details, and deletion of documents.
"""

from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Query, Request
from fastapi.responses import JSONResponse

from ingestion_service.api.schemas import (
    DocumentUploadResponse,
    DocumentResponse,
    DocumentListResponse,
    DocumentListItem,
    DeleteResponse,
    ProblemDetail,
)
from ingestion_service.adapters import get_session, DocumentRepository
from ingestion_service.domain import DocumentStatus, LegalHoldViolation
from ingestion_service.services import (
    UploadService,
    UnsupportedMediaType,
    FileTooLarge,
    QuotaExceeded,
)
from ingestion_service.middleware.auth import get_current_user, TenantContext

router = APIRouter(prefix="/documents", tags=["Documents"])


def _doc_to_response(doc) -> DocumentResponse:
    """Convert domain document to response schema."""
    return DocumentResponse(
        document_id=doc.id,
        filename=doc.filename,
        content_type=doc.content_type,
        size_bytes=doc.size_bytes,
        file_category=doc.file_category.value,
        status=doc.status.value,
        language=doc.language,
        retention_policy=doc.retention_policy,
        legal_hold=doc.legal_hold,
        uploaded_by=doc.uploaded_by,
        department_id=doc.department_id,
        collection_id=doc.collection_id,
        uploaded_at=doc.uploaded_at,
        validated_at=doc.validated_at,
        processed_at=doc.processed_at,
    )


def _doc_to_list_item(doc) -> DocumentListItem:
    """Convert domain document to list item."""
    return DocumentListItem(
        document_id=doc.id,
        filename=doc.filename,
        content_type=doc.content_type,
        size_bytes=doc.size_bytes,
        file_category=doc.file_category.value,
        status=doc.status.value,
        uploaded_at=doc.uploaded_at,
    )


@router.post(
    "",
    response_model=DocumentUploadResponse,
    status_code=201,
    responses={
        413: {"model": ProblemDetail, "description": "File too large"},
        415: {"model": ProblemDetail, "description": "Unsupported media type"},
        429: {"model": ProblemDetail, "description": "Quota exceeded"},
    },
)
async def upload_document(
    request: Request,
    file: Annotated[UploadFile, File(description="File to upload")],
    context: Annotated[TenantContext, Depends(get_current_user)],
    collection_id: UUID | None = None,
) -> DocumentUploadResponse:
    """
    Upload a new document.
    
    Accepts PDF, images (PNG, JPG), audio (MP3, WAV), and video (MP4).
    """
    # Read file content
    file_content = await file.read()

    upload_service = UploadService()

    try:
        document = await upload_service.upload(
            tenant_id=context.tenant_id,
            user_id=context.user_id,
            filename=file.filename or "unnamed",
            content_type=file.content_type or "application/octet-stream",
            file_data=file_content,
            department_id=context.department_id,
            collection_id=collection_id,
            correlation_id=getattr(request.state, "correlation_id", ""),
        )

        return DocumentUploadResponse(
            document_id=document.id,
            status=document.status.value,
            filename=document.filename,
            content_type=document.content_type,
            size_bytes=document.size_bytes,
            file_category=document.file_category.value,
            uploaded_at=document.uploaded_at,
        )

    except UnsupportedMediaType as e:
        raise HTTPException(
            status_code=415,
            detail={
                "type": "urn:waqedi:error:unsupported-media-type",
                "title": "Unsupported Media Type",
                "status": 415,
                "detail": e.message,
            },
        )
    except FileTooLarge as e:
        raise HTTPException(
            status_code=413,
            detail={
                "type": "urn:waqedi:error:file-too-large",
                "title": "File Too Large",
                "status": 413,
                "detail": e.message,
            },
        )
    except QuotaExceeded as e:
        raise HTTPException(
            status_code=429,
            detail={
                "type": "urn:waqedi:error:quota-exceeded",
                "title": "Quota Exceeded",
                "status": 429,
                "detail": e.message,
            },
        )


@router.get("", response_model=DocumentListResponse)
async def list_documents(
    context: Annotated[TenantContext, Depends(get_current_user)],
    limit: int = Query(20, ge=1, le=100),
    cursor: str | None = None,
    status: str | None = Query(None, pattern="^(UPLOADED|VALIDATED|QUEUED|PROCESSING|PROCESSED|FAILED|ARCHIVED)$"),
    file_category: str | None = Query(None, pattern="^(DOCUMENT|IMAGE|AUDIO|VIDEO)$"),
    collection_id: UUID | None = None,
) -> DocumentListResponse:
    """
    List documents in the current tenant.
    
    Returns cursor-based paginated results.
    """
    async with get_session() as session:
        repo = DocumentRepository(session, context.tenant_id)

        doc_status = DocumentStatus(status) if status else None

        documents, next_cursor = await repo.list_documents(
            limit=limit,
            cursor=cursor,
            status=doc_status,
            file_category=file_category,
            collection_id=collection_id,
        )

        return DocumentListResponse(
            items=[_doc_to_list_item(d) for d in documents],
            next_cursor=next_cursor,
        )


@router.get(
    "/{document_id}",
    response_model=DocumentResponse,
    responses={404: {"model": ProblemDetail}},
)
async def get_document(
    document_id: str,
    context: Annotated[TenantContext, Depends(get_current_user)],
) -> DocumentResponse:
    """Get document details by ID."""
    async with get_session() as session:
        repo = DocumentRepository(session, context.tenant_id)
        document = await repo.get_by_id(document_id)

        if not document:
            raise HTTPException(
                status_code=404,
                detail={
                    "type": "urn:waqedi:error:document-not-found",
                    "title": "Document Not Found",
                    "status": 404,
                    "detail": f"Document {document_id} not found",
                },
            )

        return _doc_to_response(document)


@router.delete(
    "/{document_id}",
    response_model=DeleteResponse,
    status_code=202,
    responses={
        404: {"model": ProblemDetail},
        409: {"model": ProblemDetail, "description": "Legal hold prevents deletion"},
    },
)
async def delete_document(
    document_id: str,
    context: Annotated[TenantContext, Depends(get_current_user)],
) -> DeleteResponse:
    """
    Request document deletion (soft delete).
    
    Returns 202 Accepted. Actual deletion is asynchronous.
    """
    async with get_session() as session:
        repo = DocumentRepository(session, context.tenant_id)
        document = await repo.get_by_id(document_id)

        if not document:
            raise HTTPException(
                status_code=404,
                detail={
                    "type": "urn:waqedi:error:document-not-found",
                    "title": "Document Not Found",
                    "status": 404,
                    "detail": f"Document {document_id} not found",
                },
            )

        # Check legal hold
        if document.legal_hold:
            raise HTTPException(
                status_code=409,
                detail={
                    "type": "urn:waqedi:error:legal-hold",
                    "title": "Legal Hold Active",
                    "status": 409,
                    "detail": f"Document {document_id} is under legal hold",
                },
            )

        # Soft delete
        await repo.soft_delete(document_id)

        return DeleteResponse(
            document_id=document_id,
            status="DELETED",
            message="Document has been marked for deletion",
        )
